tasks/update_go.py

```diff
@@ -0,0 +1,179 @@
+import re
+from typing import Optional
+
+from invoke import Context, exceptions, task
+
+from .go import tidy_all
+from .libs.common.color import color_message
+from .modules import DEFAULT_MODULES
+from .pipeline import update_circleci_config, update_gitlab_config
+
+GO_VERSION_FILE = "./.go-version"
+
+
+# replace the given pattern with the given string in the file
+def _update_file(path: str, pattern: str, replace: str, expected_match: int = 1):
+    # newline='' keeps the file's newline character(s)
+    # meaning it keeps '\n' for most files and '\r\n' for windows specific files
+
+    with open(path, "r", newline='') as reader:
+        content = reader.read()
+
+    content, nb_match = re.subn(pattern, replace, content, flags=re.MULTILINE)
+    if nb_match != expected_match:
+        print(
+            color_message(
+                f"WARNING: {path}: '{pattern}': expected {expected_match} matches but go {nb_match}", "orange"
+            )
+        )
+
+    with open(path, "w", newline='') as writer:
+        writer.write(content)
+
+
+# returns the current go version
+def _get_repo_go_version() -> str:
+    with open(GO_VERSION_FILE, "r") as reader:
+        version = reader.read()
+    return version.strip()
+
+
+# extracts the major version from the given string
+# eg. if the string is "1.2.3", returns "1.2"
+def _get_major_version(version: str) -> str:
+    major, _, _ = version.rpartition(".")
+    return major
+
+
+def _update_go_version_file(version: str):
+    _update_file(GO_VERSION_FILE, "[.0-9]+", version)
+
+
+def _update_gdb_dockerfile(version: str):
+    path = "./tools/gdb/Dockerfile"
+    pattern = r'(https://go\.dev/dl/go)[.0-9]+(\.linux-amd64\.tar\.gz)'
+    replace = rf'\g<1>{version}\g<2>'
+    _update_file(path, pattern, replace)
+
+
+def _update_install_devenv(version: str):
+    path = "./devenv/scripts/Install-DevEnv.ps1"
+    _update_file(path, '("Installing go )[.0-9]+"', rf'\g<1>{version}"')
+    _update_file(
+        path,
+        r'(https://dl\.google\.com/go/go)[.0-9]+(\.windows-)',
+        rf'\g<1>{version}\g<2>',
+        2,
+    )
+
+
+def _update_agent_devenv(version: str):
+    path = "./docs/dev/agent_dev_env.md"
+    pattern = r"^(You must \[install Golang\]\(https://golang\.org/doc/install\) version )`[.0-9]+`"
+    replace = rf"\g<1>`{version}`"
+    _update_file(path, pattern, replace)
+
+
+def _update_task_go(version: str):
+    path = "./tasks/go.py"
+    pattern = '("go version go)[.0-9]+( linux/amd64")'
+    replace = rf'\g<1>{version}\g<2>'
+    _update_file(path, pattern, replace)
+
+
+def _update_readme(major: str):
+    path = "./README.md"
+    pattern = r'(\[Go\]\(https://golang\.org/doc/install\) )[.0-9]+( or later)'
+    replace = rf'\g<1>{major}\g<2>'
+    _update_file(path, pattern, replace)
+
+
+def _update_go_mods(major: str):
+    mod_files = [f"./{module}/go.mod" for module in DEFAULT_MODULES]
```

Comment:
Did you check this change was fine?  This repository used `go1.18` for a long time even when using go1.19 as a compiler.

Reply:
All modules were updated to 1.20 recently, and I think starting with go 1.21 we'll *have* to keep those versions updated  because it will actually mean something (currently it's just an information, it doesn't enforce anything).
So I think updating these files as part of this task would be better